Source/PINDiskCache.m

```diff
@@ -1389,20 +1421,52 @@ - (void)setWritingProtectionOption:(NSDataWritingOptions)writingProtectionOption
     NSDataWritingOptions option = NSDataWritingFileProtectionMask & writingProtectionOption;
     
     [strongSelf lock];
-    strongSelf->_writingProtectionOption = option;
+        strongSelf->_writingProtectionOption = option;
     [strongSelf unlock];
   } withPriority:PINOperationQueuePriorityHigh];
 }
 #endif
 
+- (void)lockUntilWritable
+{
+    __unused int result = pthread_mutex_lock(&_mutex);
+    NSAssert(result == 0, @"Failed to lock PINDiskCache %@. Code: %d", self, result);
+    
+    // spinlock if the disk isn't writable
+    while (_diskWritable == NO) {
+        [self unlock];
+        usleep(100);
```

Comment:
Using this technique could cause some other problems, like a lot of threads to be spawned. We should be able to rely on the lock to be released in order to let the threads proceed.

How about a separate read-write mutex, just for cache init? At initialization of the class, the writer lock would be acquired by the code doing the read from disk. Any code calling -lockUntilWritable would first acquire the read lock (which could be grabbed by many threads at the same time, as long as the write lock isn't held). Secondly it would acquire the main [self lock].

This should allow all of them to wait on the initialization, without busying the main cache lock for operations that don't depend on _diskWritable.

As an alternative, I think there is a pthread version of the condition lock or a semaphore which can be signalled.

Reply:
Interesting, I'll play around with this.

Reply:
I think the pthread condition might be the best option. I'm *guessing* performance isn't any better for read / write locks and it's far less indicative of what we're doing. Need to actually figure out how to do pthread conditions though, not a lot of great docs…